### ISSUE TITLE
Invalid cast to CompoundTerm 

### ISSUE THREAD

**Issue by PtrMan** [opennars/opennars#311]
a. :|:
10
b. :|:
10
x. :|:
10
y. :|:
10
z. :|:

1000

a. :|:
10
b. :|:
10
x. :|:
10
y. :|:
10
z. :|:

1000

x. :|:

1000

a. :|:
10
b. :|:
10
x. :|:
10
y. :|:
10
z. :|:

1000

x. :|:

5000

x. :|:
5000

x. :|:

(let it run for some time)

java.lang.ClassCastException: org.opennars.language.Term cannot be cast to org.opennars.language.CompoundTerm
    at org.opennars.inference.RuleTables.applyRuleTable (RuleTables.java:157)
    at org.opennars.inference.RuleTables.reason (RuleTables.java:108)
    at org.opennars.control.GeneralInferenceControl.fireTermlink (GeneralInferenceControl.java:116)
    at org.opennars.control.GeneralInferenceControl.fireTaskLink (GeneralInferenceControl.java:104)
    at org.opennars.control.GeneralInferenceControl.fireConcept (GeneralInferenceControl.java:78)
    at org.opennars.control.GeneralInferenceControl.selectConceptForInference (GeneralInferenceControl.java:57)
    at org.opennars.storage.Memory.cycle (Memory.java:349)
    at org.opennars.main.Nar.cycle (Nar.java:486)
    at org.opennars.main.Nar.run (Nar.java:462)
    at java.lang.Thread.run (Thread.java:844)

and

java.lang.ClassCastException: org.opennars.language.Term cannot be cast to org.opennars.language.CompoundTerm
    at org.opennars.inference.RuleTables.applyRuleTable (RuleTables.java:191)
    at org.opennars.inference.RuleTables.reason (RuleTables.java:108)
    at org.opennars.control.GeneralInferenceControl.fireTermlink (GeneralInferenceControl.java:116)
...

**Comment by patham9** [comment]
We don't have to support non-statement events in principle, but it's "cheap enough" for us to allow it so let's handle this case too, will add the missing pre-condition. Nice find!

**Comment by patham9** [comment]
fixed by PR #312

**Comment by robert-johansson** [comment]
This might be related:

```prolog
<<(*,$1,$2) --> opposite> <=> <(*,$2,$1) --> opposite>>.
<(&&,<(*,$1,#2) --> opposite>,<(*,#2,$3) --> opposite>) ==> <$1 <-> $3>>.
<(*,A1,B2) --> opposite>.
<(*,A1,C2) --> opposite>.
<B2 <-> C2>?
20000
```

Answer <B2 <-> C2>. %1.00;0.45%
 ERR ClassCastException: org.opennars.language.Similarity cannot be cast to org.opennars.language.Inheritance [org.opennars.inference.CompositionalRules.addVariableCandidates(CompositionalRules.java:794),
.....
